commit 2c17683140b4488215cd96e49ebd842a4f2cf5e8
Author: Seiichi Ikarashi <s.ikarashi@fujitsu.com>
Date:   Wed Nov 20 15:47:09 2024 -0500

    nfs(5): Update rsize/wsize options
    
    The rsize/wsize values are not multiples of 1024 but multiples of the
    system's page size or powers of 2 if < system's page size as defined
    in fs/nfs/internal.h:nfs_io_size().
    
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/utils/mount/nfs.man b/utils/mount/nfs.man
index 233a7177..eab4692a 100644
--- a/utils/mount/nfs.man
+++ b/utils/mount/nfs.man
@@ -215,15 +215,18 @@ or smaller than the
 setting. The largest read payload supported by the Linux NFS client
 is 1,048,576 bytes (one megabyte).
 .IP
-The
+The allowed
 .B rsize
-value is a positive integral multiple of 1024.
+value is a positive integral multiple of
+system's page size
+or a power of two if it is less than
+system's page size.
 Specified
 .B rsize
 values lower than 1024 are replaced with 4096; values larger than
 1048576 are replaced with 1048576. If a specified value is within the supported
-range but not a multiple of 1024, it is rounded down to the nearest
-multiple of 1024.
+range but not such an allowed value, it is rounded down to the nearest
+allowed value.
 .IP
 If an
 .B rsize
@@ -257,16 +260,19 @@ setting. The largest write payload supported by the Linux NFS client
 is 1,048,576 bytes (one megabyte).
 .IP
 Similar to
-.B rsize
-, the
+.BR rsize ,
+the allowed
 .B wsize
-value is a positive integral multiple of 1024.
+value is a positive integral multiple of
+system's page size
+or a power of two if it is less than
+system's page size.
 Specified
 .B wsize
 values lower than 1024 are replaced with 4096; values larger than
 1048576 are replaced with 1048576. If a specified value is within the supported
-range but not a multiple of 1024, it is rounded down to the nearest
-multiple of 1024.
+range but not such an allowed value, it is rounded down to the nearest
+allowed value.
 .IP
 If a
 .B wsize

commit 42e9fc2a78ce824406a34d86e4253722ba551201
Author: Scott Mayhew <smayhew@redhat.com>
Date:   Wed Nov 20 14:54:47 2024 -0500

    nfsdctl: clarify when versions can be set on the man page
    
    Attempting to make version changes while there are nfsd threads running
    fails with -EBUSY, so make note of it on the man page.
    
    Signed-off-by: Scott Mayhew <smayhew@redhat.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/utils/nfsdctl/nfsdctl.8 b/utils/nfsdctl/nfsdctl.8
index dae1863d..38dd1d30 100644
--- a/utils/nfsdctl/nfsdctl.8
+++ b/utils/nfsdctl/nfsdctl.8
@@ -2,12 +2,12 @@
 .\"     Title: nfsdctl
 .\"    Author: Jeff Layton
 .\" Generator: Asciidoctor 2.0.20
-.\"      Date: 2024-07-22
+.\"      Date: 2024-11-18
 .\"    Manual: \ \&
 .\"    Source: \ \&
 .\"  Language: English
 .\"
-.TH "NFSDCTL" "8" "2024-07-22" "\ \&" "\ \&"
+.TH "NFSDCTL" "8" "2024-11-18" "\ \&" "\ \&"
 .ie \n(.g .ds Aq \(aq
 .el       .ds Aq '
 .ss \n[.ss] 0
@@ -176,6 +176,14 @@ all minorversions for that major version.
 .fam
 .fi
 .if n .RE
+.sp
+.if n .RS 4
+.nf
+.fam C
+Note that versions can only be set when there are no nfsd threads running.
+.fam
+.fi
+.if n .RE
 .RE
 .sp
 \fBpool\-mode\fP
diff --git a/utils/nfsdctl/nfsdctl.adoc b/utils/nfsdctl/nfsdctl.adoc
index 7119b44d..79f07cf0 100644
--- a/utils/nfsdctl/nfsdctl.adoc
+++ b/utils/nfsdctl/nfsdctl.adoc
@@ -93,6 +93,8 @@ Each subcommand can also accept its own set of options and arguments. The
   The minorversion field is optional. If not given, it will disable or enable
   all minorversions for that major version.
 
+  Note that versions can only be set when there are no nfsd threads running.
+
 *pool-mode*::
 
   Get/set the host's pool mode. This will cause the server to start threads

commit bd7f3539d1e7ed7bf2bf53bd47a7a3ff2b208306
Author: Scott Mayhew <smayhew@redhat.com>
Date:   Wed Nov 20 14:46:56 2024 -0500

    nfsdctl: fix up the help text in version_usage()
    
    The help text in version_usage() has examples with a 'v' character in
    the version string, but the format string in the sscanf() call in
    version_func() doesn't contain a 'v' character.
    
    Reviewed-by: Jeff Layton <jlayton@kernel.org>
    Signed-off-by: Scott Mayhew <smayhew@redhat.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/utils/nfsdctl/nfsdctl.c b/utils/nfsdctl/nfsdctl.c
index f7c27632..ef917ff0 100644
--- a/utils/nfsdctl/nfsdctl.c
+++ b/utils/nfsdctl/nfsdctl.c
@@ -764,9 +764,9 @@ static void version_usage(void)
 	printf("    Display currently enabled and disabled versions:\n");
 	printf("        version\n");
 	printf("    Disable NFSv4.0:\n");
-	printf("        version -v4.0\n");
+	printf("        version -4.0\n");
 	printf("    Enable v4.1, v4.2, disable v2, v3 and v4.0:\n");
-	printf("        version -2 -3 -v4.0 +4.1 +v4.2\n");
+	printf("        version -2 -3 -4.0 +4.1 +4.2\n");
 }
 
 static int version_func(struct nl_sock *sock, int argc, char ** argv)

commit 9f7a91b51ffcd8c1c3eec2909648e53c68c15e0d
Author: Benjamin Coddington <bcodding@redhat.com>
Date:   Wed Nov 20 14:42:04 2024 -0500

    libnsm: safer atomic filenames
    
    We've gotten a report of reboot notifications being sent to domains that
    end in '.new', which can happen if the NSM temporary pathname code leaves a
    file behind.  Let's fix this up by prepending a single '.' to the temp path
    which will never be resolvable as a DNS record.
    
    https://lore.kernel.org/linux-nfs/04D30B5A-C53E-4920-ADCB-C77F5577669E@oracle.com/T/#t
    
    Reported-by: Philip Rowlands <linux-nfs@dimebar.com>
    Reviewed-by: Chuck Lever <chuck.lever@oracle.com>
    Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/support/nsm/file.c b/support/nsm/file.c
index f5b44801..e0804136 100644
--- a/support/nsm/file.c
+++ b/support/nsm/file.c
@@ -184,10 +184,10 @@ static char *
 nsm_make_temp_pathname(const char *pathname)
 {
 	size_t size;
-	char *path;
+	char *path, *base;
 	int len;
 
-	size = strlen(pathname) + sizeof(".new") + 2;
+	size = strlen(pathname) + sizeof(".new") + 3;
 	if (size > PATH_MAX)
 		return NULL;
 
@@ -195,7 +195,11 @@ nsm_make_temp_pathname(const char *pathname)
 	if (path == NULL)
 		return NULL;
 
-	len = snprintf(path, size, "%s.new", pathname);
+	base = strrchr(pathname, '/');
+	strcpy(path, pathname);
+
+	len = base - pathname;
+	len += snprintf(path + len + 1, size-len, ".%s.new", base+1);
 	if (error_check(len, size)) {
 		free(path);
 		return NULL;

commit d66f18ce53b486f202f649092376f70ea87e73d8
Author: Benjamin Coddington <bcodding@redhat.com>
Date:   Wed Nov 20 14:36:21 2024 -0500

    nfs-utils: fixup statd testing simulator host arg
    
    The getopt setup for the host arg was not expecing a value, update it as
    expected
    
    Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/tests/nsm_client/nsm_client.c b/tests/nsm_client/nsm_client.c
index 8dc05917..b757209c 100644
--- a/tests/nsm_client/nsm_client.c
+++ b/tests/nsm_client/nsm_client.c
@@ -72,7 +72,7 @@ static struct timeval retrans_interval =
 static struct option longopts[] =
 {
 	{ "help", 0, 0, 'h' },
-	{ "host", 0, 0, 'H' },
+	{ "host", 1, 0, 'H' },
 	{ "name", 1, 0, 'n' },
 	{ "program", 1, 0, 'P' },
 	{ "version", 1, 0, 'v' },
@@ -136,7 +136,7 @@ main(int argc, char **argv)
 	my_name[0] = '\0';
 	host[0] = '\0';
 
-	while ((arg = getopt_long(argc, argv, "hHn:P:v:", longopts,
+	while ((arg = getopt_long(argc, argv, "hH:n:P:v:", longopts,
 				  NULL)) != EOF) {
 		switch (arg) {
 		case 'H':

commit 38b46cb1f28737069d7887b5ccf7001ba4a4ff59
Author: Robert Yang <liezhi.yang@windriver.com>
Date:   Sat Nov 9 13:37:17 2024 -0500

    reexport.h: Include unistd.h to compile with musl
    
    Signed-off-by: Robert Yang <liezhi.yang@windriver.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/support/reexport/reexport.h b/support/reexport/reexport.h
index 85fd59c1..02f86844 100644
--- a/support/reexport/reexport.h
+++ b/support/reexport/reexport.h
@@ -1,6 +1,8 @@
 #ifndef REEXPORT_H
 #define REEXPORT_H
 
+#include <unistd.h>
+
 #include "nfslib.h"
 
 enum {

commit 385c89c64274c75e65f7ce09f7cec4896b981292
Author: Scott Mayhew <smayhew@redhat.com>
Date:   Sat Nov 9 13:16:41 2024 -0500

    Makefile.am: allow mount.nfs to be writeable by owner
    
    On Red Hat-based systems, the debug symbol files are built with a
    .gdb_index section to speed up gdb initialization.  The gdb-add-index
    program calls objcopy to merge the index file into the object file.
    That fails if the object file isn't writeable by the owner.
    
    Signed-off-by: Scott Mayhew <smayhew@redhat.com>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/utils/mount/Makefile.am b/utils/mount/Makefile.am
index 5ff1148c..83a8ee1c 100644
--- a/utils/mount/Makefile.am
+++ b/utils/mount/Makefile.am
@@ -50,7 +50,7 @@ install-exec-hook:
 	  ln -sf mount.nfs mount.nfs4 && \
 	  ln -sf mount.nfs umount.nfs && \
 	  ln -sf mount.nfs umount.nfs4 && \
-	  chmod 4511 mount.nfs )
+	  chmod 4711 mount.nfs )
 uninstall-hook:
 	(cd $(DESTDIR)$(sbindir) && \
 	    rm -f mount.nfs4 umount.nfs umount.nfs4)

commit 6bfb5f050a6dc9da9a62c6c0049acc4153a645f5
Author: Seiichi Ikarashi <s.ikarashi@fujitsu.com>
Date:   Sat Nov 9 13:13:34 2024 -0500

    mount.nfs: retry NFSv3 mount after NFSv4 failure in auto negotiation
    
    The problem happens when a v3 mount fails with ETIMEDOUT after
    the v4 mount failed with EPROTONOSUPPORT, in mount auto negotiation.
    It immediately breaks from the "for" loop in nfsmount_fg()
    or nfsmount_child() due to EPROTONOSUPPORT, never doing the expected
    retries until timeout.
    
    Reviewed-by: Scott Mayhew <smayhew@redhat.com>
    Signed-off-by: Seiichi Ikarashi <s.ikarashi@fujitsu.com>
    
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff --git a/utils/mount/stropts.c b/utils/mount/stropts.c
index 2d5fa1f2..23f0a8c0 100644
--- a/utils/mount/stropts.c
+++ b/utils/mount/stropts.c
@@ -981,7 +981,7 @@ fall_back:
 	if ((result = nfs_try_mount_v3v2(mi, FALSE)))
 		return result;
 
-	if (errno != EBUSY && errno != EACCES)
+	if (errno != EBUSY && errno != EACCES && errno != ETIMEDOUT)
 		errno = olderrno;
 
 	return result;
